File: libs/vgc/core/tests/test_stringutil.py
#!/usr/bin/python3

# Copyright 2018 The VGC Developers
# See the COPYRIGHT file at the top-level directory of this distribution
# and at https://github.com/vgc/vgc/blob/master/COPYRIGHT
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import unittest

from vgc.core import (
    ParseError,
    RangeError,
    TimeUnit,
    secondsToString,
    toDoubleApprox
)

class TestRead(unittest.TestCase):

    def assertDoubleApproxIsAccurate(self, l):
        expected = [float(s) for s in l]
        result = [toDoubleApprox(s) for s in l]
        self.assertEqual(expected, result)

    def assertDoubleApproxIsApprox(self, l):
        expected = [float(s) for s in l]
        result = [toDoubleApprox(s) for s in l]
        for a,b in zip(expected, result):
            m = min(abs(a), abs(b))
            if m > 0.0:
                self.assertLess((a-b)/m, 1.0e-15)
            else:
                # Both a and b are exactly zero. This is guaranteed to pass but let's test anyway.
                self.assertTrue(a == b)

    def assertDoubleApproxParseError(self, l):
        for s in l:
            with self.assertRaises(ParseError):
                toDoubleApprox(s)

    def assertDoubleApproxRangeError(self, l):
        for s in l:
            with self.assertRaises(RangeError):
                toDoubleApprox(s)

    def testDoubleApprox(self):
        # Zero must be read accurately.
        self.assertDoubleApproxIsAccurate(["0", "0.0", ".0", "0.", "00", "0000", "00.", ".00"])
        self.assertDoubleApproxIsAccurate(["+0", "+0.0", "+.0", "+0.", "+00", "+0000", "+00.", "+.00"])
        self.assertDoubleApproxIsAccurate(["-0", "-0.0", "-.0", "-0.", "-00", "-0000", "-00.", "-.00"])
        self.assertDoubleApproxIsAccurate(["0e0", "0.0e0", ".0e0", "0.e0", "00e0", "0000e0", "00.e0", ".00e0"])
        self.assertDoubleApproxIsAccurate(["0e+0", "0.0e+0", ".0e+0", "0.e+0", "00e+0", "0000e+0", "00.e+0", ".00e+0"])
        self.assertDoubleApproxIsAccurate(["0e-0", "0.0e-0", ".0e-0", "0.e-0", "00e-0", "0000e-0", "00.e-0", ".00e-0"])

        # Integers up to 15 digits should be read accurately.
        self.assertDoubleApproxIsAccurate(["1", "2", "3", "4", "5", "6", "7", "8", "9"])
        self.assertDoubleApproxIsAccurate(["+1", "+2", "+3", "+4", "+5", "+6", "+7", "+8", "+9"])
        self.assertDoubleApproxIsAccurate(["-1", "-2", "-3", "-4", "-5", "-6", "-7", "-8", "-9"])
        self.assertDoubleApproxIsAccurate(["1.0", "2.0", "3.0", "4.0", "5.0", "6.0", "7.0", "8.0", "9.0"])
        self.assertDoubleApproxIsAccurate(["+1.0", "+2.0", "+3.0", "+4.0", "+5.0", "+6.0", "+7.0", "+8.0", "+9.0"])
        self.assertDoubleApproxIsAccurate(["-1.0", "-2.0", "-3.0", "-4.0", "-5.0", "-6.0", "-7.0", "-8.0", "-9.0"])
        self.assertDoubleApproxIsAccurate(["01", "02", "03", "04", "05", "06", "07", "08", "09"])
        self.assertDoubleApproxIsAccurate(["010", "020", "030", "040", "050", "060", "070", "080", "090"])
        self.assertDoubleApproxIsAccurate(["1e42", "1.e42", "1.0e0", "0.1e1", "0.3e1", ".1e1", "0.1234e+4", "123000e-3"])
        self.assertDoubleApproxIsAccurate(["123456789012345", "1234567890123e+2", "1.23456789012345e+15"])

        # XXX This is a bug. It should be read accurately.
        # The issue is that the 16 trailing zeros create rounding errors.
        # We'll fix later. We need to defer multiplying by 10 when reading trailing zeros until
        # we get a non-zero digit.
        self.assertDoubleApproxIsAccurate(["999999999999998"])     # 15 digits: ok
        self.assertDoubleApproxIsAccurate(["999999999999998.0"])   # 15 digits + 1 trailing zero: ok
        #self.assertDoubleApproxIsAccurate(["999999999999998.00"]) # 15 digits + 2 trailing zeros: should be ok but doesn't pass
        self.assertDoubleApproxIsApprox(["999999999999998.00"])    # Small comfort: at least this passes

        # Non-integers with finite base-2 fractional part should be read accurately.
        self.assertDoubleApproxIsAccurate(["0.5", "0.25", "0.125"])

        # Integers with more than 15 digits can only be read approximately.
        self.assertDoubleApproxIsApprox(["1234567890123456", "12345678901234567", "123456789012345678"])

        # Non-integers with infinite base-2 fractional part can only be read approximately
        self.assertDoubleApproxIsApprox(["0.01", "0.009e10", "0.3", "-0.2", "-42.55", "42.142857", "1.42E1", "42e-1"])

        # This is the smallest allowed value without underflow. It can only be read approximately.
        self.assertDoubleApproxIsApprox(["1e-307", "1000000e-313"])

        # This is the largest allowed value without overflow. It can only be read approximately.
        self.assertDoubleApproxIsApprox(["9.9999999999999999e+307", "0.0000099999999999999999e+313"])

        # Testing ParseError
        # we need at least one digit in the significand
        self.assertDoubleApproxParseError(["", ".", "+", "-", "+.", "-.", "e1", ".e1", "+e1", "-e1", "+.e1", "-.e1"])
        self.assertDoubleApproxParseError(["Hi", ".Hi", "+Hi", "-Hi", "+.Hi", "-.Hi", "e1Hi", ".e1Hi", "+e1Hi", "-e1Hi", "+.e1Hi", "-.e1Hi"])
        # we need at least one digit in the exponent
        self.assertDoubleApproxParseError(["1e", "1e+", "1e-"])
        self.assertDoubleApproxParseError(["1eHi", "1e+Hi", "1e-Hi"])
        # Can't have spaces between sign and digits
        self.assertDoubleApproxParseError(["+ 1", "- 1"])
        self.assertDoubleApproxParseError(["1e+ 1", "1e- 1"])

        # Testing RangeError
        self.assertDoubleApproxRangeError(["1e308", "10e307", "0.1e309"])

        # Test underflow: These are silently rounded to zero, no error is emitted.
        # Note: we round subnormals to zero, so that we never generate subnormals
        for s in ["1e-308"]:
            self.assertEqual(toDoubleApprox(s), 0.0)

    def testTimeUnit(self):
        TimeUnit.Seconds
        TimeUnit.Milliseconds
        TimeUnit.Microseconds
        TimeUnit.Nanoseconds

    def testSecondsToString(self):
        self.assertEqual(secondsToString(1), "1s")
        self.assertEqual(secondsToString(12), "12s")

        self.assertEqual(secondsToString(12, TimeUnit.Seconds), "12s")
        self.assertEqual(secondsToString(12, TimeUnit.Milliseconds), "12000ms")
        self.assertEqual(secondsToString(12, TimeUnit.Microseconds), "12000000µs")
        self.assertEqual(secondsToString(12, TimeUnit.Nanoseconds), "12000000000ns")

        self.assertEqual(secondsToString(12, TimeUnit.Seconds, 2), "12.00s")
        self.assertEqual(secondsToString(12, TimeUnit.Milliseconds, 2), "12000.00ms")
        self.assertEqual(secondsToString(12, TimeUnit.Microseconds, 2), "12000000.00µs")
        self.assertEqual(secondsToString(12, TimeUnit.Nanoseconds, 2), "12000000000.00ns")

        self.assertEqual(secondsToString(0.0123456, TimeUnit.Seconds, 2), "0.01s")
        self.assertEqual(secondsToString(0.0123456, TimeUnit.Milliseconds, 2), "12.35ms")

if __name__ == '__main__':
    unittest.main()

File: libs/vgc/core/tests/test_stringutil.cpp
// Copyright 2020 The VGC Developers
// See the COPYRIGHT file at the top-level directory of this distribution
// and at https://github.com/vgc/vgc/blob/master/COPYRIGHT
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <gtest/gtest.h>
#include <sstream>
#include <vgc/core/stringutil.h>

// For benchmarking
// TODO: Use Google Benchmark and separate benchmarks from unit tests
#include <vgc/core/stopwatch.h>

TEST(TestStringUtil, WriteChar) {
    std::string s;
    vgc::core::StringWriter sw(s);
    sw << 'a';
    vgc::core::write(sw, 'b');
    EXPECT_EQ(s, "ab");
}

TEST(TestStringUtil, WriteCString) {
    std::string s;
    vgc::core::StringWriter sw(s);
    sw << "Hello";
    vgc::core::write(sw, " World!");
    EXPECT_EQ(s, "Hello World!");
}

TEST(TestStringUtil, WriteInt8) {
    signed char   c = 'A';
    unsigned char d = 'A';
    vgc::Int8     i = 65;
    vgc::UInt8    j = 65;

    std::ostringstream oss;
    oss << c << d << i << j;
    EXPECT_EQ(oss.str(), "AAAA");

    std::string s;
    vgc::core::StringWriter sw(s);
    sw << c << d << i << j;
    EXPECT_EQ(s, "65656565");
}

template<typename T>
void testWriteInteger(T x) {
    std::string s;
    vgc::core::StringWriter sw(s);
    sw << x;
    EXPECT_EQ(s, std::to_string(x));
}

template<typename T>
void testWriteIntegers() {
    testWriteInteger<T>(0);
    testWriteInteger<T>(vgc::internal::type_min<T>::value);
    testWriteInteger<T>(vgc::internal::type_max<T>::value);
}

TEST(TestStringUtil, WriteIntegers) {
    testWriteIntegers<vgc::Int>();
    testWriteIntegers<vgc::Int8>();
    testWriteIntegers<vgc::Int16>();
    testWriteIntegers<vgc::Int32>();
    testWriteIntegers<vgc::Int64>();
    testWriteIntegers<vgc::UInt>();
    testWriteIntegers<vgc::UInt8>();
    testWriteIntegers<vgc::UInt16>();
    testWriteIntegers<vgc::UInt32>();
    testWriteIntegers<vgc::UInt64>();
}

TEST(TestStringUtil, WriteMixed) {
    vgc::Int x = 42;
    std::string s;
    vgc::core::StringWriter sw(s);
    sw << "The value of x is: " << x << "\n";
    EXPECT_EQ(s, "The value of x is: 42\n");
}

TEST(TestStringUtil, Benchmark) {
    vgc::core::Stopwatch t;
    int n = 1000000;
    std::vector<int> v;
    v.reserve(n);
    for (int i = 0; i < n ; ++i) {
        v.push_back(i);
    }

    t.restart();
    std::string s1a;
    s1a.reserve(6*n);
    vgc::core::StringWriter sw1a(s1a);
    for (int x : v) {
        sw1a << x;
    }
    double t1a = t.elapsed();

    t.restart();
    std::string s1b;
    vgc::core::StringWriter sw1b(s1b);
    for (int x : v) {
        sw1b << x;
    }
    double t1b = t.elapsed();

    t.restart();
    std::ostringstream oss;
    for (int x : v) {
        oss << x;
    }
    std::string s2 = oss.str();
    double t2 = t.elapsed();

    t.restart();
    std::string s3a;
    s3a.reserve(6*n);
    for (int x : v) {
        s3a += vgc::core::toString(x);
    }
    double t3a = t.elapsed();

    t.restart();
    std::string s3b;
    for (int x : v) {
        s3b += vgc::core::toString(x);
    }
    double t3b = t.elapsed();

    t.restart();
    std::string s4a;
    s4a.reserve(6*n);
    for (int x : v) {
        s4a += std::to_string(x);
    }
    double t4a = t.elapsed();

    t.restart();
    std::string s4b;
    for (int x : v) {
        s4b += std::to_string(x);
    }
    double t4b = t.elapsed();

    // Check that we get the same result
    EXPECT_EQ(s1a, s1b);
    EXPECT_EQ(s1a, s2);
    EXPECT_EQ(s1a, s3a);
    EXPECT_EQ(s1a, s3b);
    EXPECT_EQ(s1a, s4a);
    EXPECT_EQ(s1a, s4b);

    // Print timings. These normally don't show up if the test
    // succeeds, but you can manually run the test binary.
    auto unit = vgc::core::TimeUnit::Microseconds;
    std::cout << "StringWriter (reserved) ........... " << std::setw(10) << vgc::core::secondsToString(t1a, unit) << "\n";
    std::cout << "StringWriter (not reserved) ....... " << std::setw(10) << vgc::core::secondsToString(t1b, unit) << "\n";
    std::cout << "std::ostringstream (N/A) .......... " << std::setw(10) << vgc::core::secondsToString(t2, unit)  << "\n";
    std::cout << "toString(x) (reserved) ............ " << std::setw(10) << vgc::core::secondsToString(t3a, unit) << "\n";
    std::cout << "toString(x) (not reserved) ........ " << std::setw(10) << vgc::core::secondsToString(t3b, unit) << "\n";
    std::cout << "std::to_string(x) (reserved) ...... " << std::setw(10) << vgc::core::secondsToString(t4a, unit) << "\n";
    std::cout << "std::to_string(x) (not reserved) .. " << std::setw(10) << vgc::core::secondsToString(t4b, unit) << "\n";

    // Example output (Intel i7-7700K, 32GM RAM, Ubuntu 18.04, gcc 7.4.0 with -O3 -DNDEBUG):
    //
    // StringWriter (reserved) ...........   11085µs
    // StringWriter (not reserved) .......   13456µs
    // std::ostringstream (N/A) ..........   30572µs
    // toString(x) (reserved) ............   14672µs
    // toString(x) (not reserved) ........   16127µs
    // std::to_string(x) (reserved) ......   56987µs
    // std::to_string(x) (not reserved) ..   60112µs
}

int main(int argc, char **argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}
